### ISSUE TITLE
Why ShardedDDP is slower than OSS?

### ISSUE THREAD

**Issue by vycezhong** [facebookresearch/fairscale#800]
## ❓ Questions and Help

I run [the benchmark code](https://github.com/facebookresearch/fairscale/blob/master/benchmarks/oss.py) on a machine with 4 A100 GPUs and got the following results.

Optimizer | Median Throughput (img/s) (rank 0) | Peak Memory (MB)
-- | -- | --
Vanilla | 1795.73 +- 11.79 | 1473.8
OSS + DDP | 1903.18 +- 25.52 | 1209.1
OSS + ShardedDDP | 1636.11 +- 7.92 | 927.6

From [the document](https://fairscale.readthedocs.io/en/latest/deep_dive/oss_sdp_fsdp.html), OSS uses all-reduce + all-gather and ShardedDPP uses reduce + all-gather. Let the number of parameters be \phi. The communication volume of OSS and ShardedDPP is 3\phi and 2\phi, respectively. How could OSS be faster than ShardedDDP?

**Comment by anj-s** [comment]
This is something we's noticed as well in with larger number of GPUs (around the 32 GPU mark). 
@blefaudeux any thoughts on why this is the case?

**Comment by blefaudeux** [comment]
It's a bucketing + sharding artifact, in that the sharding is not order aware, and the bucketing is constrained by the sharding. Basically in that case ShardedDDP waits for the buckets to fill in, and since the sharding is not order-aware some (most) of them will wait until the end of the BW pass. It has nothing to do with the communication volume, more about overlapping.

You can try a couple of different bucketing sizes to improve the situation, else the solution would be to make the sharding order aware (there were a couple of PRs doing that (https://github.com/facebookresearch/fairscale/pull/598 and https://github.com/facebookresearch/fairscale/pull/599), but abandoned because FSDP is a superset)

**Comment by blefaudeux** [comment]
Another point is that DDP (and OSS alone) handle the grad reduction directly through C++, not Python, but it's not obvious that this would change much in terms of perfs.

**Comment by anj-s** [comment]
@blefaudeux DDP does a similar bucketing strategy though it is probably more optimized. I wonder if we can use something similar (i.e what DDP uses) as a hook and maybe we can bridge the perf gap. 

In terms of sharding not being order aware, is it that we don't shard in the order of params that we see? can you explain why using the list of params in the reverse order is not a good strategy?

**Comment by blefaudeux** [comment]
> @blefaudeux DDP does a similar bucketing strategy though it is probably more optimized. I wonder if we can use something similar (i.e what DDP uses) as a hook and maybe we can bridge the perf gap.

"bucketing + sharding artifact", not "bucketing artifact". DDP does not shard, so they can bucket in the inverse order of the grad apparitions, meaning that the buckets don't have to wait on the way back. In our case there's sharding on top

> In terms of sharding not being order aware, is it that we don't shard in the order of params that we see? can you explain why using the list of params in the reverse order is not a good strategy?
It is a good strategy, it's not the strategy used right now (greedy sharding to get roughly similar buckets), the PRs above paved the way to make it order aware indeed. It's not trivial though because users are free to change the model graph anytime, but it's definitely doable

**Comment by blefaudeux** [comment]
basically the ShardedDDP featureset is included in FSDP's, hence more work on that front is not a priority I think, we could focus on FSDP instead

**Comment by vycezhong** [comment]
@blefaudeux  I got it. Thanks for you explanation!
